File: golang/api/dmhy/dmhy.go
package dmhy

import (
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/gocolly/colly/v2"

	"dandanplay-resource-service/api"
	"dandanplay-resource-service/service"
	"dandanplay-resource-service/utils"
	"dandanplay-resource-service/utils/logger"
)

var Provider *api.Provider

func init() {
	Provider = &api.Provider{
		Name:      "動漫花園",
		IsEnabled: true,
		Route:     "/",
		Scraper:   &dmhy{},
	}
}

const (
	base               = "https://share.dmhy.org"
	typeAndSubgroupUrl = base + "/topics/advanced-search?team_id=0&sort_id=0&orderby="
	listUrl            = base + "/topics/list/page/{{.Page}}?keyword={{.Keyword}}&sort_id={{.Sort}}&team_id={{.Team}}&order=date-desc"
	indexUrl           = base + "/topics/list/page/{{.Realtime}}"
)

var (
	regexTypeId     = regexp.MustCompile("sort-(\\d+)")
	regexSubgroupId = regexp.MustCompile("team_id/(\\d+)")
)

type dmhy struct{}

func (d *dmhy) NewCollector() *colly.Collector {
	return service.NewCollector(service.CollectorOption{
		AllowProxy: true,
	})
}

func (d *dmhy) Type(types *api.Types) error {
	c := d.NewCollector()

	// CSS Selector
	// https://www.w3school.com.cn/cssref/css_selectors.asp
	c.OnHTML("select#AdvSearchSort option[value]", func(e *colly.HTMLElement) {
		id := utils.ParseInt(e.Attr("value"))
		types.Types = append(types.Types, api.Sort{
			Id:   id,
			Name: e.Text,
		})
	})

	return service.Visit(c, typeAndSubgroupUrl)
}

func (d *dmhy) Subgroup(subgroups *api.Subgroups) error {
	c := d.NewCollector()

	c.OnHTML("select#AdvSearchTeam option[value]", func(e *colly.HTMLElement) {
		id := utils.ParseInt(e.Attr("value"))
		subgroups.Subgroups = append(subgroups.Subgroups, api.Team{
			Id:   id,
			Name: e.Text,
		})
	})

	return service.Visit(c, typeAndSubgroupUrl)
}

func (d *dmhy) List(list *api.List, requestURL string, so *api.SearchOptions, query *api.ListQuery) error {
	c := d.NewCollector()

	var uniqueMap sync.Map
	once := false

	logger.Debugf("{{Parsed keyword}}  : '%s'", so.Keyword)
	logger.Debugf("{{Option $realtime}}: %d", so.Options.Realtime)
	logger.Debugf("{{Option $page}}    : %d", so.Options.Page)
	logger.Debugf("{{Option $limit}}   : %d", so.Options.Limit)

	c.OnHTML("div.nav_title>div.fl", func(e *colly.HTMLElement) {
		list.HasMore = e.ChildText("a") == "下一頁"
	})

	c.OnHTML("table#topic_list tbody tr", func(e *colly.HTMLElement) {
		if len(list.Resources) >= so.Options.Limit {
			return
		}

		var (
			Title        string
			SubgroupId   int
			SubgroupName string
			PageUrl      string
		)

		// When len(titleAndSubgroup) is 2, it means there is a team/subgroup for this resource.
		// When len(titleAndSubgroup) is 1, it means there is no team/subgroup.
		if titleAndSubgroup := e.ChildTexts("td:nth-child(3) a"); len(titleAndSubgroup) > 1 {
			SubgroupName = titleAndSubgroup[0]
			Title = titleAndSubgroup[1]
			SubgroupId = utils.MatchInt(regexSubgroupId, e.ChildAttr("td:nth-child(3) a[href]", "href"))
			PageUrl = e.ChildAttr("td:nth-child(3) a:nth-child(2)[href]", "href")
		} else {
			// SubgroupId = unknown["SubgroupId"].(int)
			// SubgroupName = unknown["SubgroupName"].(string)
			Title = titleAndSubgroup[0]
			PageUrl = e.ChildAttr("td:nth-child(3) a", "href")
		}

		TypeId := utils.MatchInt(regexTypeId, e.ChildAttr("td:nth-child(2) a[href]", "class"))

		if so.Options.Realtime > 0 {
			// isDuplicated
			if _, exist := uniqueMap.Load(PageUrl); exist {
				return
			}
			// isSubgroupNotMatched
			// an unknown SubgroupId (=0) also means not-matched
			if query.Team != 0 && SubgroupId != query.Team {
				return
			}
			// isTypeNotMatched
			if query.Sort != 0 && TypeId != query.Sort {
				return
			}
			// isKeywordNotMatched
			for _, word := range strings.Fields(so.Keyword) {
				if !utils.StrContains(Title, word) {
					return
				}
			}
		}
		uniqueMap.Store(PageUrl, true)

		TypeName := e.ChildText("td:nth-child(2) a[href]")
		Magnet := e.ChildAttr("td:nth-child(4) a[href]", "href")
		FileSize := e.ChildText("td:nth-child(5)")

		PublishDate, err := time.Parse("2006/01/02 15:04", e.ChildText("td:nth-child(1) span"))
		if err != nil {
			logger.Errorf("{{Failed when formatting time string.}} %v\n", err)
		}

		res := api.Resource{
			Title:        strings.TrimSpace(Title),
			TypeId:       TypeId,
			TypeName:     TypeName,
			SubgroupId:   SubgroupId,
			SubgroupName: SubgroupName,
			Magnet:       Magnet,
			PageUrl:      base + PageUrl,
			FileSize:     FileSize,
			PublishDate:  PublishDate.Format("2006-01-02 15:04:05"),
		}
		res.Fill()
		list.Resources = append(list.Resources, res)
	})

	if so.Options.Realtime > 0 {
		c.OnRequest(func(r *colly.Request) {
			if once {
				// If no return, c.Visit() will raise the 'Request' event again,
				// resulting in a circular call.
				// Although colly will internally prevent the third call via
				// c.HasVisited(), we should still avoid it.
				return
			}
			requestURLForRealtime, err := utils.Template(indexUrl, struct {
				Realtime int
			}{Realtime: so.Options.Realtime})
			if err != nil {
				return
			}
			once = true
			_ = c.Visit(requestURLForRealtime)
		})
	}

	return service.Visit(c, requestURL)
	// return service.VisitAndRun(c, requestURL, func(a ...interface{}) {
	// 	count := 0
	// 	regex := regexp.MustCompile(`view/(\d+)_`)
	// 	uniqueMap.Range(func(k, v interface{}) bool {
	// 		fmt.Printf("%s: %v\n", regex.FindStringSubmatch(k.(string))[1], v)
	// 		count++
	// 		return true
	// 	})
	// 	fmt.Printf("length: %d\n", count)
	// })
}

func (d *dmhy) ListQueryFormatter(query *api.ListQuery) string {
	result, err := utils.Template(listUrl, query)
	if err != nil {
		logger.Errorf("{{Failed when parsing query string.}} %v\n", err)
		return base + "/topics/list/page/1?keyword=" + query.Keyword
	}
	return result
}

File: rust/src/api/dmhy.rs
use chrono::{TimeZone, Utc};
use html_scraper::{ElementRef, Html, Selector};
use log::{error, trace};

use super::Scraper;
#[cfg(feature = "search-option")]
use crate::model::search_option::SearchOptions;
use crate::model::{Item, List, ListQuery, Resource, UNKNOWN_SUBGROUP_ID, UNKNOWN_TYPE_ID};
use crate::util::{ElementRefExt, ToInt};

/// Base URL of "dmhy".
const BASE: &str = "https://share.dmhy.org";
/// Maximum number of the search results returned by "dmhy".
const MAX_SIZE: usize = 80;

/// URL of the web page for extracting list.
#[inline]
fn _list_url(query: &ListQuery) -> String {
    format!(
        "https://share.dmhy.org/topics/list/page/{page}?keyword={keyword}&sort_id={sort}&team_id={team}&order=date-desc",
        page = query.page,
        keyword = query.keyword,
        sort = query.sort,
        team = query.team
    )
}

/// A `Scraper` for "share.dmhy.org".
pub(crate) struct Dmhy {
    /// Used to store a series of CSS selectors. (Parse once, easily reuse.)
    selectors: Selectors,
}

/// A set of reusable [`CSS Selector`](html_scraper::Selector)s.
struct Selectors {
    sort: Selector,
    team: Selector,
    has_more: Selector,
    resource: Selector,
    title_and_subgroup: Selector,
    type_id_and_name: Selector,
    magnet: Selector,
    file_size: Selector,
    publish_date: Selector,
}

impl Dmhy {
    /// Returns an instance of [`Dmhy`] Scraper for constructing a `Provider`.
    pub(crate) fn new() -> Option<Self> {
        trace!("create Scraper 'dmhy' with CSS selectors");
        Some(Self {
            selectors: Selectors {
                sort: css_selector("select#AdvSearchSort option[value]")?,
                team: css_selector("select#AdvSearchTeam option[value]")?,
                has_more: css_selector("div.nav_title>div.fl a:last-child")?,
                resource: css_selector("table#topic_list tbody tr")?,
                title_and_subgroup: css_selector("td:nth-child(3) a")?,
                type_id_and_name: css_selector("td:nth-child(2) a[href][class]")?,
                magnet: css_selector("td:nth-child(4) a[href]")?,
                file_size: css_selector("td:nth-child(5)")?,
                publish_date: css_selector("td:nth-child(1) span")?,
            },
        })
    }
}

impl Scraper for Dmhy {
    const TYPE_URL: &'static str =
        "https://share.dmhy.org/topics/advanced-search?team_id=0&sort_id=0&orderby=";
    const SUBGROUP_URL: &'static str = Self::TYPE_URL;

    #[cfg(not(feature = "search-option"))]
    fn list_url(query: &ListQuery) -> String {
        _list_url(query)
    }

    #[cfg(feature = "search-option")]
    fn list_urls(query: &ListQuery, search_options: &SearchOptions) -> Vec<String> {
        let mut urls = vec![_list_url(query)];
        if search_options.options.realtime > 0 {
            urls.push(format!(
                "https://share.dmhy.org/topics/list/page/{}",
                search_options.options.realtime
            ));
        }
        urls
    }

    fn extract_types(&self, html: String) -> Option<Vec<Item>> {
        select_or_default(&html, &self.selectors.sort, UNKNOWN_TYPE_ID)
    }

    fn extract_subgroups(&self, html: String) -> Option<Vec<Item>> {
        select_or_default(&html, &self.selectors.team, UNKNOWN_SUBGROUP_ID)
    }

    #[cfg(not(feature = "search-option"))]
    fn extract_list(&self, html: String) -> Option<List> {
        let mut result = List {
            has_more: false,
            resources: Vec::with_capacity(MAX_SIZE),
        };

        let document = Html::parse_document(&html);

        result.has_more = self.extract_has_more_from_document(&document);

        for element in document.select(&self.selectors.resource) {
            result
                .resources
                .push(self.extract_resource_from_element(element));
        }

        Some(result)
    }

    #[cfg(feature = "search-option")]
    fn extract_list(
        &self,
        htmls: Vec<String>,
        query: ListQuery,
        search_options: SearchOptions,
    ) -> Option<List> {
        let capacity = if search_options.options.limit >= MAX_SIZE {
            MAX_SIZE
        } else {
            search_options.options.limit
        };

        let mut result = List {
            has_more: false,
            resources: Vec::with_capacity(capacity),
        };

        let mut unique_set = std::collections::HashSet::with_capacity(capacity);

        let document = Html::parse_document(&htmls[0]);

        result.has_more = self.extract_has_more_from_document(&document);

        for element in document.select(&self.selectors.resource) {
            if result.resources.len() >= search_options.options.limit {
                break;
            }
            let resource = self.extract_resource_from_element(element);
            unique_set.insert(resource.page_url.clone().unwrap_or_default());
            result.resources.push(resource);
        }

        // Extract extra `Resource`s from HTML text, for the search-option `$realtime`.
        if search_options.options.realtime > 0 {
            let mut extra_resources = Vec::new();
            let document_realtime = Html::parse_document(&htmls[1]);

            'outer: for element in document_realtime.select(&self.selectors.resource) {
                let resource = self.extract_resource_from_element(element);

                // Continue Early! Therefore, we compare them in order of complexity:
                //   `team` or `sort` < existence <<< keyword

                // is-subgroup-not-matched
                //   tips:
                //     - an unknown subgroup_id (`None`) also means not-matched
                //     - `-1_i16 as u16 == 65535_u16`, but it's ok here.
                if query.team != 0 && resource.subgroup_id.unwrap_or_default() as u16 != query.team
                {
                    continue;
                }
                // is-type-not-matched
                if query.sort != 0 && resource.type_id.unwrap_or_default() as u16 != query.sort {
                    continue;
                }
                // is-duplicated
                if let Some(page_url) = &resource.page_url {
                    if unique_set.contains(page_url) {
                        // dbg!(page_url);
                        continue;
                    }
                }
                // is-keyword-not-matched
                if let Some(title) = &resource.title {
                    for word in search_options.keyword.split_whitespace() {
                        // dbg!(title);
                        if !crate::util::contains_insensitive(title, word) {
                            continue 'outer;
                        }
                    }
                }

                // Now we can be sure that `result.resources` doesn't contain this `resource`.
                // But it is not necessary to:
                //   unique_set.insert(resource.page_url.clone().unwrap_or_default());
                extra_resources.push(resource);
            }

            log::debug!(
                "a total of {} `$realtime` resources were found",
                extra_resources.len()
            );

            extra_resources.append(&mut result.resources);
            extra_resources.truncate(search_options.options.limit);
            result.resources = extra_resources;
        }

        Some(result)
    }
}

/// Search for `html` using the provided CSS `selector` and return the captured data.
/// If some of the elements are not found, they are given the `default` value provided.
///
/// This function will not fail actually, returning `Option` just to satisfy the signature
/// of the trait [`Scraper`].
fn select_or_default(html: &str, selector: &Selector, default: i16) -> Option<Vec<Item>> {
    let mut result = Vec::new();

    let document = Html::parse_fragment(html);
    for element in document.select(selector) {
        result.push(Item {
            id: element.value().attr("value").to_int().unwrap_or(default),
            name: element.inner_text(),
        })
    }

    Some(result)
}

impl Dmhy {
    /// Extract `List.has_more` info from a HTML `document`.
    #[inline]
    fn extract_has_more_from_document(&self, document: &Html) -> bool {
        if let Some(anchor) = document.select(&self.selectors.has_more).next() {
            // dbg!(anchor.html());
            anchor.inner_html().contains("下一頁")
        } else {
            false
        }
    }

    /// Extract `Resource` info from a single HTML `element`.
    fn extract_resource_from_element(&self, element: ElementRef) -> Resource {
        // When len is 2, it means there is a team/subgroup for this resource.
        // When len is 1, it means there is no team/subgroup.
        let mut title_and_subgroup = element.select(&self.selectors.title_and_subgroup);
        let subgroup_or_title = title_and_subgroup.next();
        let maybe_a_title = title_and_subgroup.next();

        let (title, page_url, subgroup_id, subgroup_name) = match (subgroup_or_title, maybe_a_title)
        {
            (Some(anchor_subgroup), Some(anchor_title)) => (
                Some(anchor_title.inner_text()),
                anchor_title.value().attr("href"),
                anchor_subgroup.value().attr("href").rsplit_then_to_int('/'),
                Some(anchor_subgroup.inner_text()),
            ),
            (Some(anchor_title), None) => (
                Some(anchor_title.inner_text()),
                anchor_title.value().attr("href"),
                None,
                None,
            ),
            (None, _) => (None, None, None, None),
        };

        let (type_id, type_name) = match element.select(&self.selectors.type_id_and_name).next() {
            Some(anchor_type) => (
                anchor_type
                    .value()
                    .attr("class")
                    .strip_prefix_then_to_int("sort-"),
                Some(anchor_type.inner_text()),
            ),
            None => (None, None),
        };

        // TODO: Sometimes there are too much Tracker URL (`&tr=`) in the magnet link,
        //   should we remove these redundant part? Only keep up to 50 links?
        //   e.g. try to search "The Fate of the Furious"
        let magnet = element
            .select(&self.selectors.magnet)
            .next()
            .and_then(|anchor_magnet| anchor_magnet.value().attr("href"));

        let file_size = element
            .select(&self.selectors.file_size)
            .next()
            .map(|td| td.inner_text());

        let publish_date = element
            .select(&self.selectors.publish_date)
            .next()
            .and_then(|span| {
                Utc.datetime_from_str(&span.inner_text(), "%Y/%m/%d %H:%M")
                    .ok()
            })
            .map(|date| date.format("%Y-%m-%d %H:%M:%S").to_string());

        Resource {
            title: title.map(|s| s.trim().to_string()),
            type_id,
            type_name: type_name.map(|s| s.trim().to_string()),
            subgroup_id,
            subgroup_name: subgroup_name.map(|s| s.trim().to_string()),
            magnet: magnet.map(|s| s.to_string()),
            page_url: page_url.map(|s| format!("{}{}", BASE, s)),
            file_size,
            publish_date,
        }
    }
}

/// Create a CSS selector from str.
#[inline]
fn css_selector(s: &str) -> Option<Selector> {
    match Selector::parse(s) {
        Ok(v) => Some(v),
        Err(e) => {
            error!("{:?}", e);
            None
        }
    }
}
